internal/olm/client/client.go

```diff
@@ -214,7 +216,70 @@ func (c Client) DoCSVWait(ctx context.Context, key types.NamespacedName) error {
 		return curPhase == olmapiv1alpha1.CSVPhaseSucceeded, nil
 	}
 
-	return wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	err := wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	if err != nil && errors.Is(err, context.DeadlineExceeded) {
+		if depCheckErr := c.checkDeploymentErrors(ctx, key, csv); depCheckErr != nil {
+			return fmt.Errorf("failed to run operator: %v %v", err, depCheckErr)
+		}
+	}
+	return err
+}
+
+// checkDeploymentErrors function loops through deployment specs of a given CSV, and prints reason
+// in case of failures, based on deployment condition.
+func (c Client) checkDeploymentErrors(ctx context.Context, key types.NamespacedName, csv olmapiv1alpha1.ClusterServiceVersion) error {
+	var depName string
+	depErrors := make(map[string]string)
+	dep := &appsv1.Deployment{}
+	for _, ds := range csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+		depName = ds.Spec.Template.Name
+		depKey := types.NamespacedName{
+			Namespace: key.Namespace,
+			Name:      depName,
+		}
+		if err := c.KubeClient.Get(ctx, depKey, dep); err != nil {
+			depErrors[depName] = err.Error()
+			break
+		}
+		for _, s := range dep.Status.Conditions {
+			if s.Type == appsv1.DeploymentAvailable && s.Status == "False" {
+				depErrors[depName] = s.Message
+			}
+		}
+	}
+	// check and print podErrors in case of deployment failures.
+	if len(depErrors) > 0 {
+		fmt.Printf("failed to run operator, deployment errors: %v\n", depErrors)
+		if podCheckerr := c.checkPodErrors(ctx); podCheckerr != nil {
+			return podCheckerr
+		}
+	}
+	return nil
+}
+
+// checkPodErrors loops through pods, and prints pod errors if any.
+func (c Client) checkPodErrors(ctx context.Context) error {
+	// loop through pods and return specific error message.
+	var podName string
+	podErrors := make(map[string]string)
+	podList := &coreV1.PodList{}
+	if err := c.KubeClient.List(ctx, podList); err != nil {
+		return fmt.Errorf("error getting Pods: %v", err)
+	}
+	for _, p := range podList.Items {
+		if p.Status.Phase != "Succeeded" {
+			podName = p.Name
+			for _, cs := range p.Status.ContainerStatuses {
+				if !cs.Ready {
+					podErrors[podName] = cs.State.Waiting.Message
```

Comment:
Nit: Remove podName declaration and directly use p.Name in  `podErrors[p.Name] = cs.State.Waiting.Message`

Reply:
```suggestion
					podErrors[p.Name] = cs.State.Waiting.Message
```